File: scripts/standings.ts
import fs from "fs";
import { type TrackName, type SeasonName } from "../src/types";
import { timeGrabber } from "scripts/timeGrabber";
import { calculateStandings } from "scripts/calculateStandings";

const activeSeason = "s5";

export const pointslessResults = ["DNF", "DSQ"];

const generateStandings = async () => {
  if (process.argv[2] === "add") {
    const parsedResults = await timeGrabber(
      process.argv[3] as SeasonName,
      process.argv[4] as TrackName,
      process.argv[5]
    );

    const resultsPath = `src/data/results/${process.argv[3]}.json`;
    const resultsFile = JSON.parse(fs.readFileSync(resultsPath, "utf8"));

    resultsFile[process.argv[4]] = parsedResults;

    fs.writeFileSync(resultsPath, JSON.stringify(resultsFile), { flag: "w" });
  }

  const standingsResults = calculateStandings(activeSeason);

  const standings = JSON.stringify(standingsResults[0]);
  const standingsPath = `src/data/standings/${activeSeason}.json`;

  try {
    fs.writeFileSync(standingsPath, standings, { flag: "w" });
    console.log(`${activeSeason} standings data saved to file successfully.`);
  } catch (error) {
    console.error("Error writing JSON data to file:", error);
  }

  const constructors = JSON.stringify(standingsResults[1]);

  const constructorsPath = `src/data/constructorsStandings/${activeSeason}.json`;

  try {
    fs.writeFileSync(constructorsPath, constructors, { flag: "w" });
    console.log(
      `${activeSeason} constructors data saved to file successfully.`
    );
  } catch (error) {
    console.error("Error writing JSON data to file:", error);
  }
};

if (process) {
  generateStandings();
}

File: scripts/standings.rs
use std::fs;
use std::env;
use std::process;
use crate::time_grabber::time_grabber;
use crate::calculate_standings::calculate_standings;

const ACTIVE_SEASON: &str = "s5";

pub const POINTSLESS_RESULTS: [&str; 2] = ["DNF", "DSQ"];

async fn generate_standings() {
    let args: Vec<String> = env::args().collect();

    if args.len() > 1 && args[1] == "add" {
        let parsed_results = time_grabber(
            &args[2],
            &args[3],
            &args[4]
        ).await;

        let results_path = format!("src/data/results/{}.json", args[2]);
        let results_file = fs::read_to_string(&results_path).expect("Unable to read file");
        let mut results_json: serde_json::Value = serde_json::from_str(&results_file).expect("Unable to parse JSON");

        results_json[&args[3]] = serde_json::to_value(parsed_results).expect("Unable to convert to JSON");

        fs::write(&results_path, serde_json::to_string(&results_json).expect("Unable to write JSON")).expect("Unable to write file");
    }

    let standings_results = calculate_standings(ACTIVE_SEASON);

    let standings = serde_json::to_string(&standings_results.0).expect("Unable to convert to JSON");
    let standings_path = format!("src/data/standings/{}.json", ACTIVE_SEASON);

    fs::write(&standings_path, standings).expect("Unable to write file");
    println!("{} standings data saved to file successfully.", ACTIVE_SEASON);

    let constructors = serde_json::to_string(&standings_results.1).expect("Unable to convert to JSON");
    let constructors_path = format!("src/data/constructorsStandings/{}.json", ACTIVE_SEASON);

    fs::write(&constructors_path, constructors).expect("Unable to write file");
    println!("{} constructors data saved to file successfully.", ACTIVE_SEASON);
}

#[tokio::main]
async fn main() {
    generate_standings().await;
}
